src/app

```diff
@@ -0,0 +1,199 @@
+#!/bin/sh
+# Copyright 2023 Northern.tech AS
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+set -e
+
+STATE="$1"
+FILES="$2"
+
+temp_dir="$FILES"/tmp
+
+cleanup() {
+    set +e
+    test -f "$PERSISTENT_STORE"/.rw_test && rm -f "$PERSISTENT_STORE"/.rw_test
+}
+
+trap cleanup 1 2 3 6 15
+
+CONFIG_FILE="/etc/mender/mender-app.conf"
+PERSISTENT_STORE="/data/mender-app"
+APP_MODULE_DIR="/usr/share/mender/app-modules/v1"
+
+version=""
+orchestrator=""
+platform=""
+app_sub_module=""
+artifact_name=""
+application_name=""
+
+if test -f "$CONFIG_FILE"; then
+    . "$CONFIG_FILE"
+fi
+
+if test -d "$PERSISTENT_STORE"; then
+    if touch "$PERSISTENT_STORE"/.rw_test; then
+        rm -f "$PERSISTENT_STORE"/.rw_test
+    else
+        echo "ERROR: cant write to persistent_store in $PERSISTENT_STORE"
+        exit 1
+    fi
+fi
+
+delta_cmd="xdelta3 -d -s"
+tar_decompress_cmd="tar -xzvf"
+
+parse_metadata() {
+    application_name=$(jq -r .application_name < "$1")
+    platform=$(jq -r .platform < "$1")
+    orchestrator=$(jq -r .orchestrator < "$1")
+    version=$(jq -r .version < "$1")
+    app_sub_module="${APP_MODULE_DIR}/${orchestrator}"
+    if test ! -f "${app_sub_module}"; then
+        echo "ERROR: ${app_sub_module} not found. exiting."
+        return 1
+    fi
+    if test ! -x "${app_sub_module}"; then
+        echo "ERROR: ${app_sub_module} not executable. exiting."
+        return 1
+    fi
+    if test ! -r "${app_sub_module}"; then
+        echo "ERROR: ${app_sub_module} not readable. exiting."
+        return 1
+    fi
+    artifact_name=$(jq -r .artifact_provides.artifact_name < "$2")
+
+    if test "${application_name}" = ""; then
+        echo "ERROR: application_name is required. exiting."
```

Comment:
>I think that executable is enough, actually. Why in theory do we need to have read rights?

 you cant execute something you cant read. since you wanted the exec flag check, then it made me add readable check as well.